common/functions/src/scalars/maths/log.rs

```diff
@@ -0,0 +1,97 @@
+// Copyright 2020 Datafuse Labs.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::f64::consts::E;
+use std::fmt;
+
+use common_datavalues::prelude::*;
+use common_datavalues::DataSchema;
+use common_datavalues::DataType;
+use common_exception::Result;
+
+use crate::scalars::function_factory::FunctionDescription;
+use crate::scalars::function_factory::FunctionFeatures;
+use crate::scalars::Function;
+
+#[derive(Clone)]
+pub struct LogFunction {
+    _display_name: String,
+    default_base: f64,
+}
+
+impl LogFunction {
+    pub fn try_create(_display_name: &str) -> Result<Box<dyn Function>> {
+        Ok(Box::new(Self {
+            _display_name: _display_name.to_string(),
+            default_base: E,
+        }))
+    }
+
+    pub fn desc() -> FunctionDescription {
+        FunctionDescription::creator(Box::new(Self::try_create))
+            .features(FunctionFeatures::default().deterministic())
+    }
+}
+
+impl Function for LogFunction {
+    fn name(&self) -> &str {
+        &*self._display_name
+    }
+
+    fn variadic_arguments(&self) -> Option<(usize, usize)> {
+        Some((1, 2))
+    }
+
+    fn return_type(&self, _args: &[DataType]) -> Result<DataType> {
+        Ok(DataType::Float64)
+    }
+
+    fn nullable(&self, _input_schema: &DataSchema) -> Result<bool> {
+        Ok(true)
+    }
+
+    fn eval(&self, columns: &DataColumnsWithField, _input_rows: usize) -> Result<DataColumn> {
+        let result = if columns.len() == 1 {
+            // Log(num) with default_base if one arg
+            let num_series = columns[0]
+                .column()
+                .to_minimal_array()?
+                .cast_with_type(&DataType::Float64)?;
+            num_series
+                .f64()?
+                .apply_cast_numeric(|v| v.log(self.default_base))
+        } else {
+            // Log(base, num) if two args
+            let num_series = columns[1]
+                .column()
+                .to_minimal_array()?
+                .cast_with_type(&DataType::Float64)?;
+            let base_column: &DataColumn =
+                &columns[0].column().cast_with_type(&DataType::Float64)?;
```

Comment:
see `ops/like.rs`

Reply:
I guess you mean taking a `zip()` instead of `apply_with_idx()` ?

I try replaced it with a `zip()`, thank you

Reply:
And we should combine the null bitmap.

Reply:
why the bitmap is required? q.q

in the string like scenery, i guess a bitmap can help represent the filtered rowset, here's a data tranform but do not really need filtering?

Reply:
> `bitmap` is to represent arrow's null bitmap

Int32Array:  `[Some(3), None, Some(5)]`

in arrow's memory layout, it contains int32 vec[3,0,5] and null bitmap b101 .

See `Validity bitmaps`  in https://arrow.apache.org/docs/format/Columnar.html.

Reply:
see: https://github.com/jorgecarleitao/arrow2/blob/main/src/compute/arity.rs#L138-L160

Reply:
TIL, thank you a lot, i need more learnings on arrow

Reply:
> see: https://github.com/jorgecarleitao/arrow2/blob/main/src/compute/arity.rs#L138-L160
do we have a wrapped `binary` function in the current code base? q.q

Reply:
Currently not, but we can have.